src/runtime/hexagon/hexagon_vtcm_pool.cc

```diff
@@ -0,0 +1,166 @@
+    allocations_.emplace_back(std::pair<char*, size_t>(ptr, nbytes));
+    last_free_entry->second -= nbytes;
+    // DebugDump();
+    return ptr;
+  }
+
+  std::pair<char*, size_t>& entry_to_allocate = free_.front();
+  for (auto entry : free_) {
+    if ((entry.second < entry_to_allocate.second) && (entry.second >= nbytes)) {
+      entry_to_allocate = entry;
+      if (entry_to_allocate.second == nbytes) {
+        break;
+      }
+    }
+  }
+  CHECK(entry_to_allocate.second >= nbytes) << "Not enough contiguous VTCM space to allocate";
+  char* ptr = entry_to_allocate.first;
+  allocations_.emplace(allocations_.end(), std::pair<char*, size_t>(ptr, nbytes));
+
+  for (auto it = free_.begin(); it != free_.end(); it++) {
+    if (ptr == it->first) {
+      if (it->second == nbytes) {
+        free_.erase(it);
+      } else {
+        it->first = it->first + nbytes;
+        it->second = it->second - nbytes;
+      }
+      break;
+    }
+  }
+  // DebugDump();
+  return ptr;
+}
+
+void HexagonVtcmPool::Free(void* ptr, size_t nbytes) {
+  char* ptr_to_free = static_cast<char*>(ptr);
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  bool found_allocation_entry = false;
+  for (auto it = allocations_.begin(); it != allocations_.end(); it++) {
+    if (ptr_to_free == it->first) {
+      CHECK(it->second == nbytes) << "Attempted to free a different size than was allocated";
+      allocations_.erase(it);
+      found_allocation_entry = true;
+      break;
+    }
+  }
+  CHECK(found_allocation_entry) << "Attempted to free a pointer that had not been allocated";
+
+  auto it = free_.begin();
+  for (; it != free_.end(); it++) {
+    CHECK(ptr_to_free != it->first) << "Attempting to free a pointer that was already free";
+    if (ptr_to_free < it->first) {
+      CHECK(ptr_to_free + nbytes <= it->first)
+          << "free_ is in an inconsistent state, freed block overlaps with next";
+      if (ptr_to_free + nbytes == it->first) {
+        // Make this entry bigger
+        it->first = ptr_to_free;
+        it->second += nbytes;
+      } else {
+        // Insert an entry before this
+        it = free_.emplace(it, std::pair<char*, size_t>(ptr_to_free, nbytes));
+      }
+      break;
+    }
+  }
+
+  if (it == free_.end()) {
+    // Insert an entry at the end
+    it = free_.emplace(it, std::pair<char*, size_t>(ptr_to_free, nbytes));
+  }
+
+  // Check for overlap with the previous entry
+  if (it != free_.begin()) {
+    auto it_prev = it;
+    it_prev--;
+    CHECK(it_prev->first + it_prev->second <= ptr_to_free)
+        << "free_ is in an inconsistent state, freed block overlaps with previous";
+    if (it_prev->first + it_prev->second == ptr_to_free) {
+      it_prev->second += it->second;
+      free_.erase(it);
+    }
+  }
+  // DebugDump();
+}
+
+void HexagonVtcmPool::DebugDump() {
+  LOG(INFO) << "VTCM list state";
+  for (auto it = allocations_.begin(); it != allocations_.end(); it++) {
```

Comment:
`for (auto [ptr, size] : allocations_)`..., same in the loop below.

Reply:
I switched from an iterator to for (auto entry : list_) for both loops.